aiosmtpd/tests/test_smtp.py

```diff
@@ -969,12 +1042,55 @@ def test_ehlo(self):
                 bytes(socket.getfqdn(), 'utf-8'),
                 b'SIZE 33554432',
                 b'SMTPUTF8',
-                b'AUTH LOGIN NULL PLAIN',
+                b'AUTH DENYFALSE DENYMISSING LOGIN NONE NULL PLAIN',
                 b'HELP',
             )
             for actual, expected in zip(lines, expecteds):
                 self.assertEqual(actual, expected)
 
+    def test_auth_byclient_plain(self):
+        with SMTP(*self.address) as client:
+            client.user = "gooduser"
+            client.password = "goodpass"
+            client.ehlo("example.com")
+            client.auth("plain", client.auth_plain)
+        self.assertEqual(b"gooduser", auth_peeker.login)
+        self.assertEqual(b"goodpass", auth_peeker.password)
+        self.assertEqual("PLAIN", auth_peeker.mechanism)
+
+    def test_auth_byclient_plain_notinitialok(self):
+        with SMTP(*self.address) as client:
+            client.user = "gooduser"
+            client.password = "goodpass"
+            client.ehlo("example.com")
+            client.auth("plain", client.auth_plain, initial_response_ok=False)
+        self.assertEqual(b"gooduser", auth_peeker.login)
+        self.assertEqual(b"goodpass", auth_peeker.password)
+        self.assertEqual("PLAIN", auth_peeker.mechanism)
+
+    def test_auth_byclient_login(self):
+        with SMTP(*self.address) as client:
+            client.user = "gooduser"
+            client.password = "goodpass"
+            client.ehlo("example.com")
+            client.auth("login", client.auth_login)
+        self.assertEqual(b"gooduser", auth_peeker.login)
+        self.assertEqual(b"goodpass", auth_peeker.password)
+        self.assertEqual("LOGIN", auth_peeker.mechanism)
+
+    # Mark this as expectedFailure because smtplib.SMTP implementation in Python>=3.5
+    # is buggy. See bpo-27820
+    @unittest.expectedFailure
```

Comment:
Until [**bpo-27820**](https://bugs.python.org/issue27820) is fixed, keep this `expectedFailure` line. When the bug is fixed, we'll ... figure something out.